---
title: "figure2_DAPs statistics"
output: 
  html_document:
    df_print: paged
  html_notebook:
editor_options:
  chunk_output_type: inline
---

for panel I

```{r setup, message=FALSE, warning=FALSE, echo=FALSE}
# packages
library(GenomeInfoDb); library(ensembldb)
library(BSgenome.Ggallus.ensembl.galGal6)
library(ChIPseeker); library(GenomicFeatures); library(org.Gg.eg.db)
library(ggplot2); library(reshape2); library(Signac)

# derived from: ~/Rscript/manuscriptFigure/fig3_peakCategories.Rmd
```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
anno_path <- "/scicore/home/tschoppp/GROUP/references/genomes/ENS_g6/chicken_Gg6_atac_pre/"

txdb = makeTxDbFromGFF(file= paste0(anno_path,"Gg6_extended_200819.filter.gtf"),
                       format = "gtf", organism = "Gallus gallus")

```

## consensus peak set

```{r, message=FALSE, warning=FALSE, echo=FALSE}
peaks=readRDS(paste0("/scicore/home/tschoppp/GROUP/lab/skeletalconvergence/fromMenghan/scATAC/peakcall/nonredundantpeaks_combined.rds"))

mtx.tmp=data.frame(seqnames=seqnames(peaks),
                  starts=start(peaks),ends=end(peaks),
                  ctype=peaks$peak_called_in,
                  score=peaks$score)
write.table(mtx.tmp,paste0("/scicore/home/tschoppp/GROUP/lab/skeletalconvergence/fromMenghan/scATAC/peakcall/nonredundantpeaks_combined.bed"), row.names = F,col.names = F,sep = "\t",quote = F)

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
peakAnno = annotatePeak(peak=paste0("/scicore/home/tschoppp/GROUP/lab/skeletalconvergence/fromMenghan/scATAC/peakcall/nonredundantpeaks_combined.bed"),TxDb = txdb,annoDb = "org.Gg.eg.db")
print(plotAnnoPie(peakAnno))
```

### DAPs: 1 vs rest (chondrocytes)

diffbind: pseudobulk data generated by step6_diffbind_chondrocytes.Rmd

```{r, message=FALSE, warning=FALSE, echo=TRUE}
peakanno.all=readRDS("/scicore/home/tschoppp/GROUP/lab/skeletalconvergence/fromMenghan/figure_drafts/fig2/figure2_peakStats.rds")

report=readRDS("/scicore/home/tschoppp/wang0007/scATAC/diffAnalysis/lns_diffbind_chondrocytes.rds")
```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
tmp.gr=StringToGRanges(paste0(report$`nasal-pval0.01`$seqnames,"-",report$`nasal-pval0.01`$start,"-",report$`nasal-pval0.01`$end))
peakAnnoDaps.n = annotatePeak(peak=tmp.gr,TxDb = txdb,annoDb = "org.Gg.eg.db")
print(plotAnnoPie(peakAnnoDaps.n))

tmp.gr=StringToGRanges(paste0(report$`somite-pval0.01`$seqnames,"-",report$`somite-pval0.01`$start,"-",report$`somite-pval0.01`$end))
peakAnnoDaps.s = annotatePeak(peak=tmp.gr,TxDb = txdb,annoDb = "org.Gg.eg.db")
print(plotAnnoPie(peakAnnoDaps.s))

tmp.gr=StringToGRanges(paste0(report$`limb-pval0.01`$seqnames,"-",report$`limb-pval0.01`$start,"-",report$`limb-pval0.01`$end))
peakAnnoDaps.l = annotatePeak(peak=tmp.gr,TxDb = txdb,annoDb = "org.Gg.eg.db")
print(plotAnnoPie(peakAnnoDaps.l))

```

## peak annotation

### top 500 DAPs: chondrocytes vs other mesenchyme 

```{r, message=FALSE, warning=FALSE, echo=TRUE}
# extract annotation
# convert percent to frequency
mtx.cate=data.frame(category=c(sapply(strsplit(peakAnno@anno$annotation," \\(ENSGAL"),'[',1),
                               sapply(strsplit(peakanno.all$nasal@anno$annotation," \\(ENSGAL"),'[',1),
                               sapply(strsplit(peakanno.all$somite@anno$annotation," \\(ENSGAL"),'[',1),
                               sapply(strsplit(peakanno.all$limb@anno$annotation," \\(ENSGAL"),'[',1)),
                    from=c(rep("consensus",length(peakAnno@anno$annotation)),
                           rep("nasal",length(peakanno.all$nasal@anno$annotation)),
                           rep("somite",length(peakanno.all$somite@anno$annotation)),
                           rep("limb",length(peakanno.all$limb@anno$annotation))))
table(mtx.cate$category,mtx.cate$from)

# keep the ones shared between these 4
annos=unique(mtx.cate$category)

```

```{r, message=FALSE, warning=FALSE, echo=TRUE,fig.width=8}
# plot
mtx.cate$category=factor(mtx.cate$category,
                      levels = c("Promoter (<=1kb)","Promoter (1-2kb)","Promoter (2-3kb)",
                                 "5' UTR","3' UTR","Exon","Intron",
                                 "Downstream (<1kb)","Downstream (1-2kb)","Downstream (2-3kb)","Distal Intergenic"))

rlevels=c("nasal","somite","limb")
res=matrix(rep(1,length(annos)*length(rlevels)),ncol=length(rlevels))
colnames(res)=rlevels
rownames(res)=annos
combn=table(mtx.cate$category,mtx.cate$from)

```

```{r, message=FALSE, warning=FALSE, echo=TRUE,fig.width=10,fig.height=5}
# plot
#get the percentage
combn.pct=apply(combn,2,function(x){x/sum(x)})
combn.pct=combn.pct*100

combn.mtx=melt(combn.pct)
combn.mtx$Var1=factor(combn.mtx$Var1,
                      levels = c("Promoter (<=1kb)","Promoter (1-2kb)","Promoter (2-3kb)",
                                 "5' UTR","3' UTR","Exon","Intron",
                                 "Downstream (<1kb)","Downstream (1-2kb)","Downstream (2-3kb)","Distal Intergenic"))
combn.mtx$Var2=factor(combn.mtx$Var2,levels = c("consensus",rlevels))

p1=ggplot(combn.mtx,
       aes(x = Var1,
           y = value,
           fill = Var2)) +
  geom_bar(width=.5, stat = "identity",
           position=position_dodge(0.7)) +
  scale_fill_manual("origin", values = c("consensus"="grey","nasal" = "#DC0078", "somite" = "#008FD4", "limb" = "#FFD01F")) +
  theme_classic() + ylab("percentage %") + xlab("category") +
  theme(axis.text.x = element_text(angle = 45, size=10, hjust = 1))
print(p1)

#add asterisk, based on adjusted p-value
# https://statisticsglobe.com/ggsignif-package-r
# *p.adj<0.05, **p.adj<0.01, ***p.adj<0.001,
p2 = p1 +
  geom_point(data = combn.mtx[combn.mtx$Var1=="Promoter (<=1kb)",],
             aes(x=Var1, y=c(10)), 
             shape = "*", size=3, show.legend = FALSE)
print(p2)

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
pdf("/scicore/home/tschoppp/GROUP/lab/skeletalconvergence/fromMenghan/figure_drafts/fig2/figure2_peakStats.pdf",height = 5,width = 12)
p2;p1
dev.off()
```

```{r, message=FALSE, warning=FALSE, echo=TRUE}

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
sessionInfo()
```
