---
title: "scATAC_cluster_annotation"
output: 
  html_document:
    df_print: paged
  html_notebook:
editor_options:
  chunk_output_type: inline
--- 

```{r setup, message=FALSE, warning=FALSE, echo=FALSE}
library(Signac); library(Seurat); library(dplyr)
library(GenomeInfoDb); library(ensembldb); library(GenomicRanges)
library(ggplot2); library(patchwork); library(reshape2)
library(sva); library(pheatmap); library(scales)
library(nnls);library(glmnet)
source("~/Rscript/testGit/scATAC/integration_function.R")

set.seed(1234)
```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
samples=list()
samples[["limb"]]=c(1,2);samples[["nasal"]]=c(3,4);samples[["somite"]]=c(5,6)
sample.name <- c("L21","L24","N15","N18","S12","S15")
data.name <- c("10x_atac_12012021","10x_atac_10042020","10x_atac_12012021","10x_atac_10042020","10x_atac_12012021","10x_atac_12012021")
anno_path <- "/scicore/home/tschoppp/GROUP/references/genomes/ENS_g6/chicken_Gg6_atac_pre/"
rna_path <- "/scicore/home/tschoppp/GROUP/lab/Chris.shared/for.Menghan"
rds_path <- paste0("~/scATAC/integration/",paste0(names(samples),"_atac_integrated010721.rds"))
scrna.name = c("forelimb","frontonasal","somites")

gene.info=read.table("~/scRNA/ensembl97_geneinfo.txt", sep="\t",header = TRUE,row.names = 1,stringsAsFactors = FALSE)
gene.info[gene.info$genename=="","genename"]=rownames(gene.info[gene.info$genename=="",])
gene.info$genename=toupper(gene.info$genename)

my.W = read.table("~/scRNA/ensembl97_wgenes.txt",sep="\t",header = TRUE,row.names = 1,stringsAsFactors = FALSE)

db <- ensDbFromGtf(gtf=paste0(anno_path,"Gg6_extended_200819.filter.gtf"), organism = "Gallus_gallus", genomeVersion = "GRCg6a", version = 97)
edb <- EnsDb(db); rm(db)
gene.ranges <- genes(edb); length(gene.ranges$gene_id)
annotations <- gene.ranges[gene.ranges$gene_biotype == 'protein_coding', ]
geneinfo = data.frame(gene_id=gene.ranges$gene_id,gene_name=gene.ranges$gene_name)

library(BSgenome.Ggallus.ensembl.galGal6)
genome=seqlengths(BSgenome.Ggallus.ensembl.galGal6)
genome=genome[names(genome) %in% c(1:33,"Z")]

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
for(i in 1:length(samples)){
  tmp.samples=names(samples)[i]
  tmp.sampleid=samples[[tmp.samples]]
  tmp.sampname=sample.name[tmp.sampleid]
  
  if(tmp.samples!="somite"){
    reduction.used='harmony'
  }else{
    reduction.used='lsi_bin'
  }
  
  # read in signac project
  hm.integrated=readRDS(rds_path[i])
  
  ## read in gene activity & scRNA
  genematrix1.raw =read.table(gzfile(paste0("~/scATAC/clustering/",tmp.sampname[1],"_ArchR_geneActivity.txt.gz")), header = TRUE, row.names = 1, sep = "\t")
  genematrix2.raw =read.table(gzfile(paste0("~/scATAC/clustering/",tmp.sampname[2],"_ArchR_geneActivity.txt.gz")), header = TRUE, row.names = 1, sep = "\t")
  colnames(genematrix1.raw)=gsub(paste0(tmp.sampname[1],"."),paste0(tmp.sampname[1],"_"),colnames(genematrix1.raw));colnames(genematrix1.raw)=gsub("\\.","-",colnames(genematrix1.raw))
  colnames(genematrix2.raw)=gsub(paste0(tmp.sampname[2],"."),paste0(tmp.sampname[2],"_"),colnames(genematrix2.raw));colnames(genematrix2.raw)=gsub("\\.","-",colnames(genematrix2.raw))
  genematrix2.raw=genematrix2.raw[rownames(genematrix1.raw),]
  genematrix.raw=cbind(genematrix1.raw,genematrix2.raw)
  #if some cells are not same, remove from seurat object
  tmp=rownames(hm.integrated@meta.data)
  flt=tmp[!(tmp %in% colnames(genematrix.raw))];length(flt)
  hm.integrated=hm.integrated[,!(colnames(hm.integrated) %in% flt)]
  # add gene activity assay
  genematrix = genematrix.raw[,rownames(hm.integrated@meta.data)]
  tmp = geneinfo; tmp = tmp[!duplicated(tmp$gene_name),]; rownames(tmp)=tmp$gene_name
  genematrix = genematrix[rownames(genematrix) %in% tmp$gene_name,]
  rownames(genematrix)=tmp[rownames(genematrix),"gene_id"]; genematrix=as.matrix(genematrix)
  genematrix = as(log(genematrix+1), "dgCMatrix")
  hm.integrated[['activity']] <- CreateAssayObject(data = genematrix)
  # use combat to remove batch effect
  edata=as.matrix(GetAssayData(hm.integrated,assay = "activity", slot = "data"))
  expressed=rowSums(edata); expressed=names(expressed[expressed>0])
  edata=edata[expressed,]
  combat_edata <- ComBat(dat=edata, 
                         batch=factor(hm.integrated$dataset), 
                         mod=NULL, par.prior=TRUE, prior.plots=FALSE)
  hm.integrated[["activity_combat"]] = CreateAssayObject(data = combat_edata)
  
  # load scRNA data
  mtx.fl_rna = readRDS(paste0(rna_path,"/",scrna.name[i],"_seurat_290420.rds"))
  my.meta=readRDS(paste0("~/scRNA/clustering/meta_",tmp.samples,".rds"))
  mtx.fl_rna=AddMetaData(mtx.fl_rna,metadata = my.meta[rownames(mtx.fl_rna@meta.data),])
  
  ##################
  ## method1: label transfer
  used=Idents(mtx.fl_rna)
  rcolor=c("grey",hue_pal()(length(levels(used))))
  names(rcolor)=c("NA",levels(used))
  DefaultAssay(hm.integrated) <- 'bins'
  thres <- 0.5
  Idents(hm.integrated)=hm.integrated$broad;Idents(mtx.fl_rna)=mtx.fl_rna$annotation_broad
  hm.integrated@meta.data$cls=factor(Idents(hm.integrated), levels=c(1:max(as.numeric(Idents(hm.integrated)))))
  prediction=getpredict(atac=hm.integrated,rna=mtx.fl_rna,usedatac = 'activity_combat', usedrna = 'integrated',embed.transfer=reduction.used)
  
  # plot
  hist(prediction$prediction.score.max,n=100) #,main="leiden 0.2"
  abline(v=0.5, col="red", lty=2, lwd=3)
  
  rcolor=c("grey",hue_pal()(length(levels(mtx.fl_rna$annotation_broad))))
  names(rcolor)=c("NA",levels(mtx.fl_rna$annotation_broad))
  stackedbar(input=prediction,thres=0.5,method="ArchR",rcolor = rcolor)
  
  # show the predicted id on UMAP
  hm.integrated <- AddMetaData(hm.integrated, metadata = prediction)
  hm.integrated@meta.data[hm.integrated@meta.data$prediction.score.max<0.5,"predicted.id"]="NA"
  DimPlot(hm.integrated, reduction = "umap",group.by = "predicted.id",cols = rcolor)
  
  # annotation for each cluster
  table(prediction$prediction.score.max > thres)
  hm.integrated <- AddMetaData(object = hm.integrated, metadata = prediction$predicted.id,col.name = "predictedId")
  hm.integrated <- AddMetaData(object = hm.integrated, metadata = prediction$prediction.score.max,col.name = "predictionScoreMax")
  predict <- hm.integrated@meta.data[hm.integrated@meta.data$predictionScoreMax > thres,]
  table(hm.integrated@meta.data$cls)*0.25; 
  table(predict$predictedId,predict$cls); table(predict$cls)*0.5
  
  ##################
  ## method2: NNLS regression
  matrix.rna=GetAssayData(mtx.fl_rna,assay = "RNA")
  matrix.atac=GetAssayData(hm.integrated, assay = "activity")
  matrix.rna=as(matrix.rna,"matrix")
  matrix.atac=as(matrix.atac,"matrix")
  
  #remove non-expressed
  gene.rna=rownames(matrix.rna[rowSums(matrix.rna>0) > 0.01*ncol(matrix.rna),]);length(gene.rna) #matrix.rna=as(matrix.rna,"matrix")
  gene.atac=rownames(matrix.atac[rowSums(matrix.atac>0) > 0.01*ncol(matrix.atac),]);length(gene.atac) #(0.005*ncol(matrix.atac))
  
  #keep only common genes
  expressed=gene.rna[gene.rna %in% gene.atac];length(expressed)
  
  #for scRNA
  DefaultAssay(mtx.fl_rna)="RNA"; Idents(mtx.fl_rna)=mtx.fl_rna$annotation_broad
  dap.rna = FindAllMarkers(object = mtx.fl_rna, min.pct = 0.1, logfc.threshold =  0.25, 
                           assay = "RNA", test.use = 'LR', #only.pos = T, 
                           latent.vars = c("CC.Difference","orig.ident"), return.thresh = 0.05,verbose = F)
  dap.rna = dap.rna[(!is.na(dap.rna$p_val) & dap.rna$p_val_adj < 0.05) & dap.rna$avg_log2FC>0,]
  dap.rna = dap.rna[order(dap.rna$cluster,-dap.rna$avg_log2FC),]
  
  dap = FindAllMarkers(object = hm.integrated, min.pct = 0.1, logfc.threshold =  0.25,
                       assay = "activity", test.use = 'LR', return.thresh = 0.05, #only.pos = T, 
                       latent.vars = c("nCount_bins","dataset"), verbose = F)
  dap = dap[!is.na(dap$p_val) & dap$p_val_adj < 0.05,]
  dap.atac=dap[dap$avg_log2FC > 0,]
  
  #get top ones
  n=100
  daptop.rna = dap.rna %>% group_by(cluster) %>% top_n(n,avg_log2FC)
  daptop.atac = dap.atac %>% group_by(cluster) %>% top_n(n,avg_log2FC)
  daptop.all=c(daptop.atac$gene,daptop.rna$gene); daptop.all=daptop.all[!duplicated(daptop.all)]
  daptop.all=daptop.all[daptop.all %in% expressed]
  dap.rna$type="scRNA";dap.atac$type="scATAC"
  daps=rbind(dap.rna[,c(1:7,11)],dap.atac)
  
  avg.rna = AverageExpression(mtx.fl_rna,assays = "RNA",features = expressed)[["RNA"]]
  avg.atac = AverageExpression(hm.integrated,assays = "activity",features = expressed)[["activity"]]
  coeff.avg=getMtpNnls(select=daptop.all,test.rna=avg.rna,test.atac=avg.atac,norm=F)
  rownames(coeff.avg)=paste0("RNA_",rownames(coeff.avg))
  pheatmap(coeff.avg, cluster_cols = F, cluster_rows = F)
  pheatmap(coeff.avg, cluster_cols = T, cluster_rows = F)
  
  # plot
  stackedbar.new(input=hm.integrated@meta.data,cname=c("dataset","annonew"))
  
  # resave
  saveRDS(hm.integrated,paste0("~/scATAC/integration/",tmp.samples,"_atac_integrated",format(Sys.Date(), "%d%m%y"),".rds"))
  
}

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
sessionInfo()
```

