---
title: "motifenrich_homer"
output: 
  html_document:
    df_print: paged
  html_notebook:
editor_options:
  chunk_output_type: inline
--- 
  
```{r setup, message=FALSE, warning=FALSE, echo=FALSE}
library(Signac); library(Seurat)

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
samples=list()
samples[["limb"]]=c(1,2);samples[["nasal"]]=c(3,4);samples[["somite"]]=c(5,6)
sample.name <- c("L21","L24","N15","N18","S12","S15")
data.name <- c("10x_atac_12012021","10x_atac_10042020","10x_atac_12012021","10x_atac_10042020","10x_atac_12012021","10x_atac_12012021")
anno_path <- "/scicore/home/tschoppp/GROUP/references/genomes/ENS_g6/chicken_Gg6_atac_pre/"
meta_path <- paste0("~/scATAC/integration/",paste0(names(samples),"_atac_integrated010721_meta.rds"))
rds_path <- paste0("~/scATAC/integration/",paste0(names(samples),"_atac_integrated010721.rds"))

cls.mesenchyme=list(); cls.nonmesenchyme=list()
cls.mesenchyme[["limb"]]=c("1","2","3","4","5","6","8","9","12");cls.mesenchyme[["nasal"]]=c("2","3","4");cls.mesenchyme[["somite"]]=c("1")
cls.nonmesenchyme[["limb"]]=c("13","14","15","7");cls.nonmesenchyme[["nasal"]]=c("1","5","6","7","8","11","12","13","14","15");cls.nonmesenchyme[["somite"]]=c("2","3","4","5","6","7","8","9","10","11","12","13","14")

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
for(i in 1:length(samples)){
  ## read in data
  tmp.samples=names(samples)[i]
  
  ## generate bed file from positive DAPs
  daps.broad=read.table(paste0("~/scATAC/diffAnalysis/",tmp.samples,"_daps_broad.txt"))
  daps.broad=daps.broad[!(daps.broad$cluster %in% cls.mesenchyme[[tmp.samples]]),]
  for(j in unique(daps.broad$cluster)){
    daps.tmp=daps.broad[daps.broad$avg_log2FC>0 & daps.broad$cluster==j,]
    tmp.bed=data.frame(chr=sapply(strsplit(daps.tmp$gene,":|-"), `[`, 1),
                       start=sapply(strsplit(daps.tmp$gene,":|-"), `[`, 2),
                       end=sapply(strsplit(daps.tmp$gene,":|-"), `[`, 3))
    write.table(tmp.bed,paste0("~/scATAC/motifenrich/",tmp.samples,"_res/",tmp.samples,"_posDaps_cls",j,".txt"),
                sep = "\t",row.names = F, col.names = F,quote = F)
  }
  
  daps.fine=read.table(paste0("~/scATAC/diffAnalysis/",tmp.samples,"_daps_fine.txt"))
  for(j in unique(daps.fine$cluster)){
    daps.tmp=daps.fine[daps.fine$avg_log2FC>0 & daps.fine$cluster==j,]
    tmp.bed=data.frame(chr=sapply(strsplit(daps.tmp$gene,":|-"), `[`, 1),
                       start=sapply(strsplit(daps.tmp$gene,":|-"), `[`, 2),
                       end=sapply(strsplit(daps.tmp$gene,":|-"), `[`, 3))
    write.table(tmp.bed,paste0("~/scATAC/motifenrich/",tmp.samples,"_res/",tmp.samples,"_posDaps_cls",j,".txt"),
                sep = "\t",row.names = F, col.names = F,quote = F)
  }
  
  # for mesenchyme
  daps1=read.table(paste0("~/scATAC/diffAnalysis/",tmp.samples,"_daps_mesenchymeVSnonMesenchyme.txt"))
  daps1$gene=rownames(daps1)
  daps1=daps1[daps1$avg_log2FC>0,]
  tmp.bed1=data.frame(chr=sapply(strsplit(daps1$gene,":|-"), `[`, 1),
                      start=sapply(strsplit(daps1$gene,":|-"), `[`, 2),
                      end=sapply(strsplit(daps1$gene,":|-"), `[`, 3))
  write.table(tmp.bed1,paste0("~/scATAC/motifenrich/",tmp.samples,"_res/",tmp.samples,"_posDaps_mesenchymeVSnonMesenchyme.txt"),
              sep = "\t",row.names = F, col.names = F,quote = F)
  
  ## run homer
  # default vertebrate database, & de novo motif search
  # findMotifsGenome.pl limb_posDaps_cls7.txt Gallus_gallus.GRCg6a.dna.toplevel.fa homer_limb_motifs_cls7 -len 8,9,10,11,12,13,14,15,16,17,18,19,20,21,22  -mset vertebrates -size -250,250 -fdr 5 -p 4 -cache 4000
  test.cls=NULL; test.cls=paste0("cls",c(unique(daps.broad$cluster),unique(daps.fine$cluster)))
  test.cls=c(test.cls,"mesenchymeVSnonMesenchyme")
  for(j in test.cls){
    x="~/scATAC/motifenrich/homer_defaultVertebrates.sh"
    bed.tmp=paste0("~/scATAC/motifenrich/",tmp.samples,"_res/",tmp.samples,"_posDaps_",j,".txt")
    if(j=="mesenchymeVSnonMesenchyme"){j="clsMschym"}
    out.tmp=paste0("~/scATAC/motifenrich/",tmp.samples,"_res/","homer_",tmp.samples,"_motifs_",j)
    cmd.tmp=paste0("sbatch ",x," -b ",bed.tmp," -o ",out.tmp)
    system(cmd.tmp)
  }
  
}

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}

```

## log

```{r, message=FALSE, warning=FALSE, echo=TRUE}
sessionInfo()
```
