---
title: "scATAC_ptbinMatrix"
output: 
  html_document:
    df_print: paged
  html_notebook:
editor_options:
  chunk_output_type: inline
---

```{r setup, message=FALSE, warning=FALSE, echo=FALSE}
library(Signac);library(Seurat)
library(ggplot2);library(limma)
library(slingshot)

set.seed(1234)
```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
## !!changed the way to get ptbins for scATAC.
## !!changed average of motif activity, as default will expm1().
ptheatmap=function(scatac=subset.atac,
                   ptatac=pt.meta,ptatac.selected="pseudotime.chondrocytes",
                   nbins=80,
                   origin="limb"){
  #############
  ## scatac
  
  #add motif activity
  chromvar.mtx=read.table(gzfile(paste0("~/scATAC/motifenrich/",origin,"_chromvarDenovo-LNSmotifs.txt.gz")), sep = "\t")
  colnames(chromvar.mtx)=gsub("\\.1","-1",colnames(chromvar.mtx))
  rownames(chromvar.mtx)=substr(rownames(chromvar.mtx),1,20)
  chromvar.mtx=as.matrix(chromvar.mtx)
  rownames(chromvar.mtx)=gsub("sub-","sub",rownames(chromvar.mtx))

  subset.atac[["chromvar.name"]]=CreateAssayObject(data=chromvar.mtx[,colnames(subset.atac)])
  DefaultAssay(subset.atac) <- 'chromvar.name'
  subset.atac@assays$chromvar.name@scale.data=chromvar.mtx[,colnames(subset.atac)]
  #subset.atac=ScaleData(subset.atac)
  
  #get pseudotime & ptbins
  subset.atac <- AddMetaData(
    object = subset.atac,
    metadata = ptatac[rownames(subset.atac@meta.data),ptatac.selected],
    col.name = "pseudotime"
  )
  my.data=subset(subset.atac,cells=rownames(ptatac[!is.na(ptatac[,ptatac.selected]),]))
  my.meta=my.data@meta.data
  
  
  #put cells into n pseudotime bins
  if(origin=="limb"){
    i=1
  }else if(origin=="somite"){
    i=2
  }else if(origin=="nasal"){
    i=3
  }
  get_slshot=readRDS(my.pfiles[i])
  
  #need to use same interval as scRNA, so that can calculate Pearson correlation
  n=nbins
  ptrna=slingPseudotime(get_slshot[["xtsne"]])[,cid[i]]
  ptrna=ptrna[!is.na(ptrna)]
  max=max(ptrna) #max(my.meta$pseudotime)
  intervals=seq(0,max,by=max/n)
  my.meta$ptbins=1000
  for(j in 1:(length(intervals)-1)){
    ptmin=as.numeric(intervals[j])
    ptmax=as.numeric(intervals[(j+1)])
    my.meta[my.meta$pseudotime >= ptmin & my.meta$pseudotime < ptmax,"ptbins"]=j
  }
  my.meta[my.meta$ptbins==1000,"ptbins"]=n
  p1=ggplot(data.frame(my.meta), aes(x=ptbins)) + 
    geom_histogram(colour="black", fill="white",bins=n-1) + 
    theme_classic()+ggtitle("scATAC ptBins")+ylab("cell number")
  print(p1);print(table(my.meta$ptbins))
  
  #calculate average motif activity
  my.data=AddMetaData(my.data,my.meta[rownames(my.data@meta.data),c("pseudotime","ptbins")])
  my.data$ptbins=factor(my.data$ptbins,levels = as.character(1:n))
  Idents(my.data)=my.data$ptbins
  avg.atac=AverageExpression(my.data,assays = "chromvar.name",slot = "scale.data")[["chromvar.name"]] #!! need to use this, otherwise will be expm1()  #no log, no scaled
  colnames(avg.atac)=paste0(origin,"_bin",colnames(avg.atac))
  avg.atac=as.data.frame(avg.atac)
  avg.atac[(nrow(avg.atac)+1),]=table(my.data$ptbins)
  rownames(avg.atac)[nrow(avg.atac)]="cellnumberFORbins"
  print("finish scATAC part (average gene activity)!")
  
  #calculate average gene activity
  avg.atac.geneact=AverageExpression(my.data,assays = "activity.name")[["activity.name"]] #no log, no scaled
  colnames(avg.atac.geneact)=paste0(origin,"_bin",colnames(avg.atac.geneact))
  avg.atac.geneact=as.data.frame(avg.atac.geneact)
  print("finish scATAC part (average motif activity)!")
  
  
  #############
  ## scrna
  
  #get scRNA data
  subset.rna=readRDS(my.files[i])
  subset.rna@active.assay="RNA"
  meta.rna=readRDS(my.mfiles[i]); dim(meta.rna)
  
  #add pseudotime & subset
  meta.rna$pseudotime=slingPseudotime(get_slshot[["xtsne"]])[,cid[i]][rownames(meta.rna)]
  meta.rna=meta.rna[!is.na(meta.rna$pseudotime),]; dim(meta.rna)
  subset.rna=subset(subset.rna,cells = rownames(meta.rna))
  meta.rna=meta.rna[rownames(subset.rna@meta.data),]
  
  #put cells into n pseudotime bins
  #n=nbins
  #max=max(meta.rna$pseudotime)
  #intervals=seq(0,max,by=max/n)
  meta.rna$ptbins=1000
  for(j in 1:(length(intervals)-1)){
    ptmin=as.numeric(intervals[j])
    ptmax=as.numeric(intervals[(j+1)])
    meta.rna[meta.rna$pseudotime >= ptmin & meta.rna$pseudotime < ptmax,"ptbins"]=j
  }
  meta.rna[meta.rna$ptbins==1000,"ptbins"]=n
  p2=ggplot(data.frame(meta.rna), aes(x=ptbins)) + 
    geom_histogram(colour="black", fill="white",bins=n-1) + 
    theme_classic()+ggtitle("scRNA ptBins")+ylab("cell number")
  print(p2);print(table(meta.rna$ptbins))
  
  #calculate average expression
  subset.rna=AddMetaData(subset.rna,meta.rna[rownames(subset.rna@meta.data),c("pseudotime","ptbins")])
  subset.rna$ptbins=factor(subset.rna$ptbins,levels = as.character(1:n))
  Idents(subset.rna)=subset.rna$ptbins; assay.used="RNA"
  my.average=AverageExpression(subset.rna,assays = assay.used)[[assay.used]] #no log, no scaled
  colnames(my.average)=paste0(samples,"_bin",colnames(my.average))
  my.average=as.data.frame(my.average)
  
  print("finish scRNA part (average gene expression)!")
  
  avg=list()
  avg[["scatac"]]=avg.atac
  avg[["scrna"]]=my.average
  avg[["scatac.geneact"]]=avg.atac.geneact
  return(avg)

}

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
ptime.cls=list()
ptime.cls[["limb"]]=paste(c("WNT5AMesenchyme","chondrocytes","lateChondrocytes","amb_mesenchymeORnsConnectiveTissue"),collapse = "|")
ptime.cls[["nasal"]]=paste(c("neuralcrestDerivedMesenchyme","eyeConnectiveTissue","chondrocytes"),collapse = "|")
ptime.cls[["somite"]]=paste(c("somite2","somite1","chondrocytes"),collapse = "|")

my.origins=c("limb","somite","nasal")
my.files=paste0("~/scRNA/clustering/subset_",my.origins,".rds")
my.pfiles=paste0("~/scRNA/pseudotime/ptime_",my.origins,"_slingshot.rds")
my.mfiles=paste0("~/scRNA/clustering/meta_",my.origins,".rds")
cid=c("curve1","curve2","curve4")
```

## limb

```{r, message=FALSE, warning=FALSE, echo=TRUE}
samples="limb"
subset.atac=readRDS(paste0("/scicore/home/tschoppp/wang0007/scATAC/integration/",samples,"_atac_intgtFine010721.rds"))
my.meta=readRDS(paste0("/scicore/home/tschoppp/wang0007/scATAC/integration/",samples,"_atac_integrated010721_meta.rds"))
subset.atac=AddMetaData(subset.atac,metadata = my.meta[rownames(subset.atac@meta.data),c("broad","annotation_broad","fine","annotation_fine")],col.name = c("broad","annotation_broad","fine","annotation_fine"))
pt.meta=readRDS(paste0("~/scATAC/pseudotime/ptime_scatac_",samples,"_summary.rds"))

DimPlot(subset.atac,reduction = "xtsne",group.by = "annotation_fine",label = T)+NoLegend()
```

described in the paper:"Chromatin and gene-regulatory dynamics of the developing human cerebral cortex at single-cell resolution"
These PCCs, motif activity and gene expression, were computed across the pseudotime-pseudobulk samples.

Need to set more criteria: 1) paralogs need to be TFs as well;
2) if both original ones and paralogs pass, paralogs need to performance better than original one & statistic significant.

```{r, message=FALSE, warning=FALSE, echo=TRUE,fig.width=4,fig.height=6}
#seurat::transferdata, relaxed cutoff
ph=ptheatmap(scatac=subset.atac,
            ptatac=pt.meta,ptatac.selected="pseudotime.transferData.relaxed",
            nbins=80,
            origin=samples)

dim(ph$scatac) #
dim(ph$scrna)
```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
# the pseudotime bins should be comparable
# save all 1240 motifs
# save all genes with ensembl id

# save ph file
saveRDS(ph,paste0("~/scATAC/motifReassign/ptbinMatrix_",samples,".rds"))
```

## nasal

```{r, message=FALSE, warning=FALSE, echo=TRUE}
samples="nasal"
subset.atac=readRDS(paste0("/scicore/home/tschoppp/wang0007/scATAC/integration/",samples,"_atac_intgtFine010721.rds"))
my.meta=readRDS(paste0("/scicore/home/tschoppp/wang0007/scATAC/integration/",samples,"_atac_integrated010721_meta.rds"))
subset.atac=AddMetaData(subset.atac,metadata = my.meta[rownames(subset.atac@meta.data),c("broad","annotation_broad","fine","annotation_fine")],col.name = c("broad","annotation_broad","fine","annotation_fine"))
pt.meta=readRDS(paste0("~/scATAC/pseudotime/ptime_scatac_",samples,"_summary.rds"))

DimPlot(subset.atac,reduction = "xtsne",group.by = "annotation_fine",label = T)+NoLegend()
```

```{r, message=FALSE, warning=FALSE, echo=TRUE,fig.width=4,fig.height=6}
#seurat::transferdata, relaxed cutoff
ph=ptheatmap(scatac=subset.atac,
            ptatac=pt.meta,ptatac.selected="pseudotime.transferData.relaxed",
            nbins=80,
            origin=samples)

dim(ph$scatac) #
dim(ph$scrna)
```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
# the pseudotime bins should be comparable
# save all 1240 motifs
# save all genes with ensembl id

# save ph file
saveRDS(ph,paste0("~/scATAC/motifReassign/ptbinMatrix_",samples,".rds"))
```

## somite

```{r, message=FALSE, warning=FALSE, echo=TRUE}
samples="somite"
subset.atac=readRDS(paste0("/scicore/home/tschoppp/wang0007/scATAC/integration/",samples,"_atac_intgtFine010721.rds"))
my.meta=readRDS(paste0("/scicore/home/tschoppp/wang0007/scATAC/integration/",samples,"_atac_integrated010721_meta.rds"))
subset.atac=AddMetaData(subset.atac,metadata = my.meta[rownames(subset.atac@meta.data),c("broad","annotation_broad","fine","annotation_fine")],col.name = c("broad","annotation_broad","fine","annotation_fine"))
pt.meta=readRDS(paste0("~/scATAC/pseudotime/ptime_scatac_",samples,"_summary.rds"))

DimPlot(subset.atac,reduction = "xtsne",group.by = "annotation_fine",label = T)+NoLegend()
```

```{r, message=FALSE, warning=FALSE, echo=TRUE,fig.width=4,fig.height=6}
#seurat::transferdata, relaxed cutoff
ph=ptheatmap(scatac=subset.atac,
            ptatac=pt.meta,ptatac.selected="pseudotime.transferData.relaxed",
            nbins=80,
            origin=samples)

dim(ph$scatac) #
dim(ph$scrna)
```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
# the pseudotime bins should be comparable
# save all 1240 motifs
# save all genes with ensembl id

# save ph file
saveRDS(ph,paste0("~/scATAC/motifReassign/ptbinMatrix_",samples,".rds"))
```

## log

```{r, message=FALSE, warning=FALSE, echo=TRUE}

```

```{r, message=FALSE, warning=FALSE, echo=TRUE}
sessionInfo()
```
